### PR TITLE
Support High-DPI awareness and modal/custom window style. (windows)

### PR DESCRIPTION
**Pull request by testdrive-profiling-master**
# Support High-DPI awareness per a monitor
When setup the monitor as 4k@175% DPI.
It looks blurred.
image

Now supports High-DPI awareness as like below.
image

# Support full custom window style without default system title bar.
Above image normal window form in current state.
But now additional window style be supported by `void webui_set_window_style(size_t window, webui_window_style id, uint32_t focused_bgcolor, uint32_t unfocused_bgcolor)` function.
```cpp
typedef enum {
	WEBUI_WINDOW_STYLE_NORMAL = 0,      // 0. normal window (min/max/close + titlebar & resizable)
	WEBUI_WINDOW_STYLE_MODAL,           // 1. modal window (close + titlebar, no resizable) - currently windows only
	WEBUI_WINDOW_STYLE_CUSTOM,          // 2. custom window (no titlebar, but resizable) - currently windows only
	WEBUI_WINDOW_STYLE_CUSTOM_MODAL,    // 3. custom modal window (no titlebar, no resizable) - currently windows only
} webui_window_style;
```
### `WEBUI_WINDOW_STYLE_NORMAL`
as above image...!
### `WEBUI_WINDOW_STYLE_MODAL` (without minimize/maximize button, and not resizable.)
image

### `WEBUI_WINDOW_STYLE_CUSTOM_MODAL` (no title bar, no resizable)
This seems like it would be helpful when creating full custom web based windows.
image

Since custom style window does not have a title bar, so it provide a separate way to initiate dragging window via the `bool webui_command_window(size_t window, webui_window_command command)` function.
```cpp
typedef enum {
	WEBUI_WINDOW_COMMAND_HIDE = 0,		// 0. hide window (currently windows only)
	WEBUI_WINDOW_COMMAND_SHOW,			// 1. show window (currently windows only)
	WEBUI_WINDOW_COMMAND_START_DRAG,	// 2. start drag window (currently windows only)
	WEBUI_WINDOW_COMMAND_NORMAL,		// 3. normal window (currently windows only)
	WEBUI_WINDOW_COMMAND_MAXIMIZE,		// 4. maximize window (currently windows only)
	WEBUI_WINDOW_COMMAND_MINIMIZE,		// 5. minimize window (currently windows only)
	WEBUI_WINDOW_COMMAND_IS_NORMAL,		// 6. is normal window (currently windows only)
	WEBUI_WINDOW_COMMAND_IS_MAXIMIZED,	// 7. is maximized window (currently windows only)
	WEBUI_WINDOW_COMMAND_IS_MINIZED,	// 8. is minimized window (currently windows only)
} webui_window_command;
```

I hope someone smarter than me implements this for linux/mac.

### CODE REVIEW

**Comment by jinzhongjia** [comment]
nice feature!
You can ignore the zig test for the time being.

File: include/webui.h
Comment (mochalins) [inline include/webui.h:495]: This does not seem like the right docs for this function.

File: build.zig
Comment (mochalins) [inline build.zig:101]: Should follow indentation above

File: include/webui.h
Comment (mochalins) [inline include/webui.h:56]: Where does this Windows version limitation come from? If it's just for the DPI awareness, it seems unreasonable to limit Windows version for the entire library.

Comment (testdrive-profiling-master) [reply include/webui.h:56]: No limit on type of windows, 7,8,10,11 what ever.
Actually it just related on build version 1703, old windows will need to be updated.

Refer this(win7 need to update win7.sp1.) : https://github.com/AvaloniaUI/Avalonia/issues/2095

And current CI's zig build results of all possible windows type are passed, right?

File: include/webui.h
Comment (testdrive-profiling-master) [reply include/webui.h:495]: Good I will fix it.

File: build.zig
Comment (testdrive-profiling-master) [reply build.zig:101]: I missed it.
I use tab on my editor, github's default tab size is 8.
Can I add configuration file for github to set 'tab size 4'?

File: include/webui.h
Comment (mochalins) [reply include/webui.h:56]: The CI only contains one version of Windows. There is indeed a limit on the version of Windows, as the `SetProcessDpiAwarenessContext` function is only available on `Windows 10, version 1703` and higher. Wouldn't it be better to use the older `SetProcessDpiAwareness` to support up to `Windows 8.1`?

Comment (testdrive-profiling-master) [reply include/webui.h:56]: You're right, I misunderstood.
I fixed it up like this.
```cpp
// Windows DPI awareness support
#ifdef _WIN32
#include <shellscalingapi.h>	// for 'SetProcessDpiAwareness' function use
static void __InitDpiAwareness(void) {
	{	// minimum platform : Windows 10, version 1703 / Windows Server 2016
		BOOL (*pSetProcessDpiAwarenessContext)(DPI_AWARENESS_CONTEXT value) = (BOOL (*)(DPI_AWARENESS_CONTEXT))GetProcAddress(GetModuleHandle("user32.dll"), "SetProcessDpiAwarenessContext");

		if (pSetProcessDpiAwarenessContext) {
			pSetProcessDpiAwarenessContext(DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2);
			return;
		}
	}
	{	// minimum platform : Windows 8.1 / Windows Server 2012 R2
		HMODULE hLib = LoadLibrary("Shcore.dll");	// dynamic load lib is required.
		if (hLib) {
			HRESULT (*pSetProcessDpiAwareness)(PROCESS_DPI_AWARENESS value) = (HRESULT (*)(PROCESS_DPI_AWARENESS))GetProcAddress(hLib, "SetProcessDpiAwareness");
			if (pSetProcessDpiAwareness) {
				if (pSetProcessDpiAwareness(PROCESS_PER_MONITOR_DPI_AWARE) != S_OK) {
					pSetProcessDpiAwareness(PROCESS_SYSTEM_DPI_AWARE);
				}
				return;
			}
		}
	}
	{	// minimum platform : Windows Vista / Windows Server 2008
		BOOL (*pSetProcessDPIAware)() = (BOOL (*)())GetProcAddress(GetModuleHandle("user32.dll"), "SetProcessDPIAware");
		if (pSetProcessDPIAware) {
			pSetProcessDPIAware();
		}
	}
	
	// If it reached to here, what ancient Windows is used?
}

static UINT __GetDpiForWindow(HWND hwnd) {
	static bool bInitialized	= false;
	static UINT (*pGetDpiForWindow)(HWND hwnd) = NULL;
	
	if (!bInitialized) {
		pGetDpiForWindow = (UINT (*)(HWND))GetProcAddress(GetModuleHandle("user32.dll"), "GetDpiForWindow");
		bInitialized = true;
	}
	// if it's supported
	if(pGetDpiForWindow) return pGetDpiForWindow(hwnd);
	// or not, return default DPI value. (under a Win 10.)
	return USER_DEFAULT_SCREEN_DPI;
}
#endif
```
Same problem with `GetDpiForWindows()` is fixed.

Comment (mochalins) [reply include/webui.h:56]: The change looks great! Thank you for your work.

**Review by mochalins — APPROVED** [review]

**Comment by mochalins** [comment]
> Is there is a reason why this PR did not get merged?

Quite frankly, I'm not sure. Looked mergeable to me, but I've been waiting on core maintainer feedback for any PRs that change/add-to/remove-from public API. Perhaps just slow review turnaround time from core maintainers.

**Pull request closed without merging**
